proc/threads.c

```diff
@@ -1497,13 +1499,56 @@ int threads_sigpost(process_t *process, thread_t *thread, int sig)
 
 	hal_cpuReschedule(&threads_common.spinlock, &sc);
 
-	if (kill)
-		proc_kill(process);
-
 	return EOK;
 }
 
 
+static int _threads_checkSignal(thread_t *selected, process_t *proc, cpu_context_t *signalCtx)
+{
+	unsigned int sig;
+
+	sig = (selected->sigpend | proc->sigpend) & ~selected->sigmask;
+	if ((sig != 0) && (proc->sighandler != NULL)) {
+		sig = hal_cpuGetLastBit(sig);
+
+		if (hal_cpuPushSignal(selected->kstack + selected->kstacksz, proc->sighandler, signalCtx, sig) == 0) {
+			selected->sigpend &= ~(1 << sig);
+			proc->sigpend &= ~(1 << sig);
+			return 0;
+		}
+	}
+
+	return -1;
+}
+
+
+void threads_setupUserReturn(ptr_t retval)
+{
+	spinlock_ctx_t sc;
+	cpu_context_t *ctx, *signalCtx;
+	void *f;
+	void *kstackTop;
+	thread_t *thread;
+
+	hal_spinlockSet(&threads_common.spinlock, &sc);
+	thread = _proc_current();
+
+	kstackTop = thread->kstack + thread->kstacksz;
+	ctx = kstackTop - sizeof(*ctx);
+	signalCtx = (void *)((char *)hal_cpuGetUserSP(ctx) - sizeof(*signalCtx));
+
+	if (_threads_checkSignal(thread, thread->process, signalCtx) == 0) {
```

Comment:
Shouldn't we check for `thread->process == NULL`?

Reply:
Can we have userspace thread without a process? I'll add check if needed.